mypyc/genops.py

```diff
@@ -987,14 +1028,7 @@ def gen_arg_default(self) -> None:
                         "Unsupported default argument")
                 target = self.environment.lookup(arg.variable)
                 assert isinstance(target, AssignmentTargetRegister)
-                error_block, body_block = BasicBlock(), BasicBlock()
-                self.add(Branch(target.register, error_block, body_block, Branch.IS_ERROR))
-                self.activate_block(error_block)
-                reg = self.accept(arg.initializer)
-                self.add(Assign(target.register,
-                                self.coerce(reg, target.register.type, arg.initializer.line)))
-                self.goto(body_block)
-                self.activate_block(body_block)
+                self.assign_if_null(target, self.accept(arg.initializer), arg.initializer.line)
```

Comment:
This changes the behavior of default arguments slightly: the initializer is now evaluated unconditionally, instead of only if the argument wasn't passed in. It's not a huge deal but we still shouldn't do it, I think.

I think the way to fix this while preserving a nice general API for the `assign_if_null` function is to have it take a `Callable[[], Value]` instead of a value, so that it can arrange for the initializer to be evaluated only in the `error_block`.